lib/vrl/compiler/src/type_def.rs

```diff
@@ -926,6 +952,197 @@ impl TypeDef {
+                                    .iter()
+                                    .map(|(idx, kindinfo)| match idx {
+                                        Index::Index(idx)
+                                            if *index > 0 && *idx == *index as usize =>
+                                        {
+                                            (Index::Index(*idx), newkind.clone())
+                                        }
+                                        _ => {
+                                            // If we don't have a specific index into the array to
+                                            // update, we need to merge this type in since index 1
+                                            // could contain one type and index 2 could contain another,
+                                            // the whole array will need to be a union of the two.
+                                            if !indexed {
+                                                (
+                                                    *idx,
+                                                    kindinfo.clone().merge(
+                                                        newkind.clone(),
+                                                        false,
+                                                        true,
+                                                    ),
+                                                )
+                                            } else {
+                                                (*idx, kindinfo.clone())
+                                            }
+                                        }
+                                    })
+                                    .collect(),
+                            )
+                        }
+
+                        (kind, _, _) => kind.clone(),
+                    })
+                    .collect(),
+            ),
+        }
+    }
+
+    /// Updates the type definition at the given path with the provided kindinfo.
+    fn update_path(&self, path: &LookupBuf, kind: &KindInfo) -> Self {
+        let segments = path.as_segments();
+        let peekable = segments.iter().peekable();
+        let kind = Self::update_segment(&self.kind, peekable, kind);
+
+        TypeDef {
+            fallible: self.fallible,
+            kind,
+        }
+    }
+
+    /// Assuming path points at an Array, this will take the typedefs for that array,
+    /// And will remove it returning a set of it's options.
+    ///
+    /// For example the typedef for this object:
+    /// `{ "nonk" => { "shnoog" => [ { "noog" => 2 }, { "noog" => 3 } ] } }`
+    ///
+    /// Is converted to a typedef for this array:
+    /// `[ { "nonk" => { "shnoog" => { "noog" => 2 } } },
+    ///    { "nonk" => { "shnoog" => { "noog" => 3 } } },
+    ///  ]`
+    ///
+    /// This is used for the `unnest` vrl function.
+    ///
+    pub fn invert_array_at_path(self, path: &LookupBuf) -> Self {
```

Comment:
This seems like a highly specialized function that will only ever be useful for the `unnest` function, unless you have some other examples for which this might be useful?

The reason I'm asking, is that it might make sense to move this logic to the function type def impl itself because (a) the name of this function isn't very declarative as-is, and (2) it removes the need to add another public function to `TypeDef`.

It might make more sense to have `update_path` become public, have it accept a `TypeDef` instead of `KindInfo` (to allow one to also set the fallibility at a path), and then copy this function to the `unnest` impl and have it call `update_path`.

Reply:
Yes, that makes sense.

Reply:
I do think this is better, but I have a query [here](https://github.com/timberio/vector/pull/7404/files#r666891231) for you to consider..